support enum values in ElasticSearch

VinnyQ opened the pull request:
WHAT:  Add search capability for blueflood so that users can pass in the param "include_enum_values" and be able to retrieve enum values along with the metric names in the result when hitting the /v2.0/:tenantId/metrics/search endpoint.
(see story:  https://jira.rax.io/browse/CMD-741;  Note -- I do not think we should implement the enum version of the DiscoveryWriter class in this story, but should move the writing of that class to the rollup story where the new index will actually be created during rollup).

HOW:
- extract ElasticIO.Discovery inner class to standalone Discovery class
- extract ElasticIO.ESFieldLabel enum to standalone enum class
- add configs for enums elasticsearch module
- add EnumElasticIO class with insertDiscovery() and search() methods.
- modify SearchResult to store enum values field
- modify HttpMetricsIndexHandler to check for include_enum_values
- modify HttpMetricsIndexHandler to include enum values in json result

Please refer to the new "enums" index mapping for the index and its fields to be created by metrics rollup of enum values:  https://github.com/VinnyQ/blueflood-chef/blob/add_elasticsearch_mappings_for_enums/site-cookbooks/old-elasticsearch/files/default/enums.json

NOTE:  This deploy will need changes to blueflood.conf (after the code is merged)
To add the following property:
```
ENUMS_DISCOVERY_MODULES=com.rackspacecloud.blueflood.io.EnumElasticIO
```

Do this in https://github.com/rackerlabs/blueflood-chef.  See blueflood-chef PR https://github.com/rackerlabs/blueflood-chef/pull/635 which does the following:
1.  add the enums.json mappings file for elasticsearch templates
2.  add the ENUMS_DISCOVERY_MODULES property to blueflood.conf

And also, for now we will have to *manually* create another ElasticSearch alias "metric_metadata_read" that also points to the new "enums" index.  The code to do so is:

```
curl -XPOST 'http://localhost:9200/_aliases' -d '
{
    "actions" : [
        { "add" : { "index" : "enums", "alias" : "metric_metadata_read" } }
    ]
}'
```

VinnyQ commented on blueflood-core/src/main/java/com/rackspacecloud/blueflood/io/SearchResult.java:
```diff
@@ -1,37 +1,74 @@
+
         return result;
     }
 
```
Need to add test for .equals() to verify values compare in arraylist

VinnyQ replied:
tests added.

VinnyQ commented on blueflood-core/src/main/java/com/rackspacecloud/blueflood/types/EnumMetric.java:
```diff
@@ -0,0 +1,137 @@
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+
```
Note:  we might be able to refactor both this class and the Metric class to both extend from a BaseMetric class that implements IMetric, so that they can share some common code such as setTTL and getTTL.  

Let's decide whether we should do this now or later when we are actually implementing the writing the data to cassandra story.

VinnyQ replied:
Or conversely, we can just reuse the Metric class and modify it to also support EnumValues if passed in, like so:

https://gist.github.com/VinnyQ/28f6431340d8b2397ffb

And not have this new EnumMetric class.

tilogaat replied:
@VinnyQ I think we can add a field in the BluefloodEnumRollup class like this. 

```
HashSet<String> stringEnumValues = new HashSet<String>() 
```

Then IMO, you will not need this EnumMetric class, you should just be able to use the PreaggregatedMetric class. 
Notice that PreaggregatedMetric already has a Rollup field. 
https://github.com/rackerlabs/blueflood/blob/10ca96311e71adc913ee30458ef70d83423399f5/blueflood-core/src/main/java/com/rackspacecloud/blueflood/types/PreaggregatedMetric.java#L28

We add a method to retrieve the rollup in PreaggregatedMetric class as follows. 

```
 public Rollup getRollup() {
        return this.value;
 }
```

Then you can modify your EnumElasticIO in this way: 

```
if (obj instanceof PreaggregatedMetric ) {
                PreaggregatedMetric metric = (PreaggregatedMetric)obj;
                if (metric.getRollupType() == RollupType.ENUM) {
                    BluefloodEnumRollup enumRollup = (BluefloodEnumRollup)metric.getRollup();
                    Locator locator = metric.getLocator();
                    Discovery md = new Discovery(locator.getTenantId(), locator.getMetricName());

                    Map<String, Object> info = new HashMap<String, Object>();

                    HashSet<String> enumValues = enumRollup.getEnumValues();

                    if (enumValues != null) {
                        info.put(ESFieldLabel.enum_values.toString(), enumValues);
                    }
                }
            }
```

It makes sense to use the Rollup class to store values, because we are going to build them there anyways right? With the bad metrics table. 

I like this idea of EnumMetric and having separate Metric classes. Unfortunately, currently the metric type specific logic is only in Rollups.

VinnyQ replied:
@tilogaat great suggestion!  Please take a look at the new code here.  

I simplified your suggestions further and just added the "getStringEnumValues()" method in BluefloodEnumRollup, which returns the keySet of all enum values as an ArrayList<String>, and use the PreaggreatedMetric class instead of a new EnumMetric class to store and retrieve the metrics.

This means that we'll have to compose a batch of PreaggreatedMetric which include enum values to be called by EnumElasticIO.insertDiscovery(), instead of a batch of EnumMetrics.

I will see where to make that call when I am working on the DiscoveryWriter.

VinnyQ commented on blueflood-elasticsearch/src/main/java/com/rackspacecloud/blueflood/io/Discovery.java:
```diff
@@ -0,0 +1,76 @@
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
```
This class is extracted from ElasticIO (where it was an inner class) so that the new EnumElasticIO class can also use it.

VinnyQ commented on blueflood-elasticsearch/src/main/java/com/rackspacecloud/blueflood/io/ESFieldLabel.java:
```diff
@@ -0,0 +1,25 @@
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
```
Note:  this enum class is extracted from the ElasticIO class so that it can be used by both the existing ElasticIO class and the new EnumElasticIO class.

VinnyQ commented on blueflood-elasticsearch/src/main/java/com/rackspacecloud/blueflood/io/ElasticIO.java:
```diff
@@ -132,7 +122,7 @@ private static String getUnit(Metric metric) {
         return metric.getUnit();
     }
 
```
Oops, method below should have stayed private.  I had to made it internal because I was trying to extend the ElasticIO class for EnumElasticIO, which turned out to be really hard to set up testing for.  I will make this private again in the next commit.

VinnyQ replied:
Nevermind, i am going to leave it as internal so we can unit test it by itself.  Adding test for it in ElasticIOTest, which should also be in the com.rackspacecloud.blueflood.io package (and not service).

VinnyQ replied:
tests added.

VinnyQ commented on blueflood-elasticsearch/src/main/java/com/rackspacecloud/blueflood/io/EnumElasticIO.java:
```diff
@@ -0,0 +1,209 @@
+import java.util.*;
+
+import static org.elasticsearch.index.query.QueryBuilders.*;
+
```
This is the new class for creating and querying enum indexes.  Note that this solution depends on adding another "metric_metadata_read" alias to point to the new "enums" index (this alias is currently pointing to both the "metric_metadata" and "metric_metadata_v2" indexes), so that when we query the "metric_metadata_read" alias, it will find all metrics in all 3 indexes (metric_metadata, metric_metadata_v2, enums) with the metric names matching the query value.

The dedupResults() method will handle "include_enum_values" by giving results with enum_values fields precedent over those without, when adding to the results hashmap  (with unique key by metric_name) .

Note to @GeorgeJahad, you are actually right here in that we *don't* have to use a new alias for the enums metrics.  Another way to do it is to do two separate search, one against the read alias (metric_metadata_read), and once against the new "enums" index, then combined/union the results into a new hashmap with the results from "enums" taking precedent.  If you would prefer we do it this way instead, please let me know.

tilogaat replied:
@VinnyQ Am I right in assuming that EnumElasticIO will be used in Rollups to write values? I did not see any code currently to integrate with the ingest pipeline.

VinnyQ replied:
Yes, and it will be called by a new class that's similar to DiscoveryWriter.  I was going to work on that class and the integration in story:  https://jira.rax.io/browse/CMD-749

VinnyQ commented:
@GeorgeJahad @tilogaat @goru97 Please review when you have a chance.

VinnyQ commented on blueflood-elasticsearch/src/main/java/com/rackspacecloud/blueflood/io/ESFieldLabel.java:
```diff
@@ -0,0 +1,25 @@
+    unit,
+    enum_values
+}
+
```
Note: this ESFieldLabel enum class is extracted from the ElasticIO class so that it can be used by both the existing ElasticIO class and the new EnumElasticIO class.

VinnyQ commented on blueflood-elasticsearch/src/main/java/com/rackspacecloud/blueflood/io/Discovery.java:
```diff
@@ -0,0 +1,76 @@
+        this.tenantId = tenantId;
+        this.metricName = metricName;
+    }
+
```
This Discovery class is extracted from ElasticIO (where it was an inner class) so that the new EnumElasticIO class can also use it.

VinnyQ commented:
Hi @tilogaat @GeorgeJahad @goru97, please start code review again.  I have discussed with tilo and worked with her to reuse her BluefloodEnumRollup class, eliminating the need to create a new EnumMetric class implementing IMetric to store the metric data inc. enum values (we switched to using the PreaggregatedMetric class which is also used by rollups for the same purpose).

I am happy with this code, and uncertain whether to we actually need a writer class akin to DiscoveryWriter at the moment.  I'll look at what's needed when doing the "Check for bad enum metrics" story:  https://jira.rax.io/browse/CMD-749.

tilogaat commented:
@VinnyQ can you make a list of all the chef changes for this story?

tilogaat commented on blueflood-core/src/main/java/com/rackspacecloud/blueflood/io/SearchResult.java:
```diff
@@ -1,37 +1,87 @@
+
+        s += "]";
+        return s;
     }
```
NIT: Can we use StringBuilder or String.format

VinnyQ replied:
Fixed!  using StringBuilder!

tilogaat commented:
@VinnyQ Can you add an integration test that will explain how it all comes together?
Check out 
```
HttpAnnotationsEndToEndTest
```
for inspiration. Let me know if I missed a similar test though.

tilogaat commented on blueflood-elasticsearch/src/test/java/com/rackspacecloud/blueflood/io/EnumElasticIOTest.java:
```diff
@@ -0,0 +1,309 @@
+        Assert.assertTrue("results2 should contain tenant2 entry1", results2.contains(tenant2Entry1));
+        Assert.assertTrue("results2 should contain tenant2 entry2", results2.contains(tenant2Entry2));
+        Assert.assertTrue("results2 should contain tenant2 entry3", results2.contains(tenant2Entry3));
+        Assert.assertTrue("results1 should not contain tenant1 entry1", !results2.contains(tenant1Entry1));
```
NIT: should this be results2 should not contain tenant1 entry1?

VinnyQ replied:
good catch, fixed!

GeorgeJahad commented on blueflood-elasticsearch/src/main/java/com/rackspacecloud/blueflood/io/EnumElasticIO.java:
```diff
@@ -0,0 +1,215 @@
+                    info.put(ESFieldLabel.enum_values.toString(), enumValues);
+                }
+
+                md.withAnnotation(info);
```
what are annotations and why are we putting the enum_values there?

VinnyQ replied:
annotation are actually "extra fields" that are not either the metric_name or tenant_id in the Discovery class.  The way the Discovery class is setup is that any fields you would like to include in the index to be written to ES (such as "unit"), you would add it as annotation.

See the code here in the extracted Discovery class (formerly and inner class of ElasticIO), in method createSourceContent(), to see how the annotations is retrieved to prepare the json source for the index:  https://github.com/VinnyQ/blueflood/blob/3d6d289712cfcfb84187a5a069632aad08ba8300/blueflood-elasticsearch/src/main/java/com/rackspacecloud/blueflood/io/Discovery.java#L70-L72

VinnyQ replied:
So I do agree with your implicit comment here @GeorgeJahad ;) "annotation" is probably a bad name for field in the Discovery class.  We can change this to .withFields or .withSourceFields if you like.  That might be a better name for this.

GeorgeJahad replied:
I would rename it.

VinnyQ replied:
property and method renamed to .fields and .getSourceFields().

GeorgeJahad commented on blueflood-elasticsearch/src/test/java/com/rackspacecloud/blueflood/io/EnumElasticIOTest.java:
```diff
@@ -0,0 +1,309 @@
+        verify(indexRequestBuilderMock).setRouting(TENANT_1);
+        verify(indexRequestBuilderMock).setRouting(TENANT_2);
+    }
+
```
I like your thoroughness here, but I wonder if we aren't coupling the test too tightly with the implementation?  What do you think?

VinnyQ replied:
We might be.  If the tests started failing because of implementation changes and becoming tedious to maintain, we can just remove these routing and  create call and just assert the locator field.  Or do you want me to remove it now?

GeorgeJahad replied:
I wouldn't change it now, and in general I like how thorough your tests are.  In this particular case, it just seems like it's a little too strongly coupled to the implementation, (and I would guard against that in the future.)

GeorgeJahad commented:
+1

VinnyQ commented:
@tilogaat re:  end to end integration test similar to HttpAnnotationsEndToEndTest, this is not immediately implementable because the enum_metrics are not ingested and added to ES as indexes similar to other metrics.  There's a rollup step in between.  

The closest I have right now is the EnumElasticIOTest search tests (https://github.com/VinnyQ/blueflood/blob/3d6d289712cfcfb84187a5a069632aad08ba8300/blueflood-elasticsearch/src/test/java/com/rackspacecloud/blueflood/io/EnumElasticIOTest.java) where I would create the ES indexes using insertDiscovery() and the search for the results in the test.

However, I am planning or adding the integration end to end test from the rollups creation to the index searching once I added that feature in a different story.  Once the rollup feature is added to create the ES enums indexes during rollup, I'll have an end to end story that will

1.  post/ingest enum metrics
2.  run the rollups
3.  check that the indexes are created in ES

Sounds good?

tilogaat commented:
+1 Sounds good vinny

VinnyQ commented:
@GeorgeJahad @tilogaat all issues from code review addressed.  Do you guys want to take a look at it one more time before I merge it?
